File: app/src/main/java/com/example/mvvmtest/ui/activity/ChatActivity.java
package com.example.mvvmtest.ui.activity;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.text.TextUtils;
import android.widget.Button;
import android.widget.EditText;

import com.example.mvvmtest.R;
import com.example.mvvmtest.adapter.ChatAdapter;
import com.example.mvvmtest.dagger.component.ApiController;
import com.example.mvvmtest.manager.FlechPreferences;
import com.example.mvvmtest.model.Message;
import com.example.mvvmtest.util.Constant;
import com.example.mvvmtest.viewmodel.ChatViewModel;
import com.github.nkzawa.socketio.client.Socket;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ChatActivity extends AppCompatActivity {

    private ChatViewModel chatViewModel;
    private ChatAdapter chatAdapter;
    private int nickname;

    @Inject
    protected Socket socket;

    @Inject
    protected FlechPreferences sharedFlechPreferences;

    @BindView(R.id.chatRecyclerView)
    protected RecyclerView recyclerView;

    @BindView(R.id.sendButton)
    protected Button sendButton;

    @BindView(R.id.sendEditText)
    protected EditText sendEditText;

    @BindView(R.id.toolbar)
    protected Toolbar toolbar;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        ButterKnife.bind(this);
        ApiController.getAppComponent().inject(this);
        nickname = getIntent().getIntExtra("nickname", Constant.NICKNAME);
        initViewModel();
        initSocket();
        initAdapter();
        checkChatStatus();

    }

    private void checkChatStatus() {
        if (isConnected()) {
            toolbar.setSubtitle("Connected");
        } else {
            toolbar.setSubtitle("Disconnected");
        }
    }

    private void initAdapter() {
        //chatAdapter = new ChatAdapter(ChatActivity.this, chatViewModel.getMessages().getValue());
        recyclerView.setAdapter(chatAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(ChatActivity.this));
    }

    private void initViewModel() {
        chatViewModel = ViewModelProviders.of(this).get(ChatViewModel.class);
       /* chatViewModel.init(nickname);
        chatViewModel.getMessages().observe(this, new Observer<List<Message>>() {
            @Override
            public void onChanged(@Nullable List<Message> messages) {
                chatAdapter.notifyDataSetChanged();
            }
        });
        chatViewModel.getIsTyping().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isTyping) {
                if (isTyping) {
                    toolbar.setSubtitle("Escribiendo...");
                } else {
                    checkChatStatus();
                }
            }
        });*/

    }

    private void initSocket() {
        socket.connect();
        //socket.on(Constant.SocketEvent.NEW_MESSAGE, chatViewModel.onNewMessage);
        //socket.on(Constant.SocketEvent.TYPING, chatViewModel.onTyping);
    }

    @OnClick(R.id.sendButton)
    protected void sendMessage() {
        String text = sendEditText.getText().toString().trim();
        if (TextUtils.isEmpty(text)) {
            return;
        }
        //chatViewModel.sendMessage(socket, new Message(Constant.SENDER, nickname, text));
        sendEditText.setText("");
    }

    public boolean isConnected() {
        return socket.connected();
    }

    private void connect() {
        socket.connect();
    }

    private void disconnect() {
        socket.off(Constant.SocketEvent.NEW_MESSAGE);
        socket.off(Constant.SocketEvent.TYPING);
        socket.disconnect();
    }

}

File: app/src/main/java/com/example/mvvmtest/ui/activity/ChatActivity.kt
package com.example.mvvmtest.ui.activity

import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View

import com.example.mvvmtest.R
import com.example.mvvmtest.adapter.ChatAdapter
import com.example.mvvmtest.dagger.component.ApiController
import com.example.mvvmtest.manager.FlechPreferences
import com.example.mvvmtest.model.Message
import com.example.mvvmtest.model.Typing
import com.example.mvvmtest.util.*
import com.example.mvvmtest.viewmodel.ChatViewModel
import com.github.nkzawa.socketio.client.Socket

import javax.inject.Inject

import kotlinx.android.synthetic.main.content_chat.*

class ChatActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var chatViewModel: ChatViewModel
    private lateinit var chatAdapter: ChatAdapter
    private lateinit var layoutManager: LinearLayoutManager
    private var nickname: Int? = null

    @Inject
    lateinit var socket: Socket

    @Inject
    lateinit var flechPreferences: FlechPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)
        ApiController.getAppComponent().inject(this)
        //nickname = intent.getIntExtra("nickname", Constant.NICKNAME)
        nickname = Constant.NICKNAME
        initViewModel()
        initAdapter()
        initViews()
        initSocket()
        setupEditTextBehavior()
    }


    private fun setupEditTextBehavior() {
        sendEditText.afterTextChanged {
            if (it.isEmpty()) sendTyping(false) else sendTyping(true)
        }
    }

    private fun initViews() {
        sendButton.setOnClickListener(this)
    }

    private fun initAdapter() {
        layoutManager = LinearLayoutManager(this@ChatActivity)
        layoutManager.stackFromEnd = true
        chatRecyclerView.layoutManager = layoutManager
        val mutableList: MutableList<Message> = arrayListOf()
        chatAdapter = ChatAdapter(mutableList)
        chatRecyclerView.adapter = chatAdapter

        chatRecyclerView.onLayoutChange {
            if (it) scrollToBottom()
        }

    }

    private fun scrollToBottom() {
        layoutManager.smoothScrollToPosition(chatRecyclerView, null, chatAdapter.itemCount)
    }

    private fun initViewModel() {

        val viewModelFactory = ChatViewModelFactory(nickname!!)
        chatViewModel = ViewModelProviders.of(this@ChatActivity, viewModelFactory)
                .get(ChatViewModel::class.java)

        chatViewModel.getMessages()
        chatViewModel.messagesLiveData.observe(this, Observer {
            Log.d(Constant.TAG, "entro en observer " + it.toString())
            chatAdapter.addMessages(it)
            scrollToBottom()
        })

        chatViewModel.typingLiveData.observe(this, Observer {
            performTyping(it)
        })

    }

    private fun performTyping(typing: Boolean) {
        Log.d(Constant.TAG, "entro en performTyping $typing")
        if (typing) chatLinearLayout.visibility = View.VISIBLE else chatLinearLayout.visibility = View.GONE
    }

    private fun initSocket() {
        socket.connect()
        socket.on(Constant.SocketEvent.NEW_MESSAGE, chatViewModel.onNewMessageListener)
        socket.on(Constant.SocketEvent.TYPING, chatViewModel.onTypingListener)

    }

    private fun sendMessage() {
        sendTyping(false)
        val text = sendEditText!!.text.toString().trim { it <= ' ' }
        if (TextUtils.isEmpty(text)) {
            return
        }
        chatViewModel.sendMessage(socket, Message(Constant.SENDER, nickname!!, text))
        sendEditText.setText("")
    }

    private fun sendTyping(typing: Boolean) {
        chatViewModel.sendTyping(socket, Typing(Constant.SENDER, Constant.NICKNAME, typing))
    }

    private fun disconnect() {
        socket.off(Constant.SocketEvent.NEW_MESSAGE)
        socket.off(Constant.SocketEvent.TYPING)
        socket.disconnect()
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.sendButton -> {
                sendMessage()
            }
            else -> {

            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        disconnect()
    }

}
